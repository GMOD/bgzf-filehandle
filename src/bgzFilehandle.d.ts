/// <reference types="node" />
import GziIndex from './gziIndex';
import { GenericFilehandle } from 'generic-filehandle';
export default class BgzFilehandle {
    filehandle: GenericFilehandle;
    gzi: GziIndex;
    constructor({ filehandle, path, gziFilehandle, gziPath, }: {
        filehandle?: GenericFilehandle;
        path?: string;
        gziFilehandle?: GenericFilehandle;
        gziPath?: string;
    });
    stat(): Promise<import("generic-filehandle").Stats & {
        size: any;
        blocks: undefined;
        blksize: undefined;
    }>;
    getUncompressedFileSize(): Promise<any>;
    _readAndUncompressBlock(blockBuffer: Buffer, [compressedPosition]: [number], [nextCompressedPosition]: [number]): Promise<Buffer>;
    read(buf: Buffer, offset: number, length: number, position: number): Promise<{
        bytesRead: number;
        buffer: Buffer;
    }>;
}
