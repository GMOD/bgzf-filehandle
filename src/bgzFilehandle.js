"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var unzip_1 = require("./unzip");
var gziIndex_1 = __importDefault(require("./gziIndex"));
var generic_filehandle_1 = require("generic-filehandle");
var BgzFilehandle = /** @class */ (function () {
    function BgzFilehandle(_a) {
        var filehandle = _a.filehandle, path = _a.path, gziFilehandle = _a.gziFilehandle, gziPath = _a.gziPath;
        if (filehandle)
            this.filehandle = filehandle;
        else if (path)
            this.filehandle = new generic_filehandle_1.LocalFile(path);
        else
            throw new TypeError('either filehandle or path must be defined');
        if (!gziFilehandle && !gziPath && !path)
            throw new TypeError('either gziFilehandle or gziPath must be defined');
        this.gzi = new gziIndex_1["default"]({
            filehandle: gziFilehandle,
            path: !gziFilehandle && !gziPath && path ? gziPath : path + ".gzi"
        });
    }
    BgzFilehandle.prototype.stat = function () {
        return __awaiter(this, void 0, void 0, function () {
            var compressedStat, _a, _b, _c;
            var _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0: return [4 /*yield*/, this.filehandle.stat()];
                    case 1:
                        compressedStat = _e.sent();
                        _b = (_a = Object).assign;
                        _c = [compressedStat];
                        _d = {};
                        return [4 /*yield*/, this.getUncompressedFileSize()];
                    case 2: return [2 /*return*/, _b.apply(_a, _c.concat([(_d.size = _e.sent(),
                                _d.blocks = undefined,
                                _d.blksize = undefined,
                                _d)]))];
                }
            });
        });
    };
    BgzFilehandle.prototype.getUncompressedFileSize = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, uncompressedPosition, size, buf, bytesRead, lastBlockUncompressedSize;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.gzi.getLastBlock()];
                    case 1:
                        _a = _b.sent(), uncompressedPosition = _a[1];
                        return [4 /*yield*/, this.filehandle.stat()];
                    case 2:
                        size = (_b.sent()).size;
                        buf = Buffer.allocUnsafe(4);
                        return [4 /*yield*/, this.filehandle.read(buf, 0, 4, size - 28 - 4)];
                    case 3:
                        bytesRead = (_b.sent()).bytesRead;
                        if (bytesRead !== 4)
                            throw new Error('read error');
                        lastBlockUncompressedSize = buf.readUInt32LE(0);
                        return [2 /*return*/, uncompressedPosition + lastBlockUncompressedSize];
                }
            });
        });
    };
    BgzFilehandle.prototype._readAndUncompressBlock = function (blockBuffer, _a, _b) {
        var compressedPosition = _a[0];
        var nextCompressedPosition = _b[0];
        return __awaiter(this, void 0, void 0, function () {
            var next, blockCompressedLength, unzippedBuffer;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        next = nextCompressedPosition;
                        if (!!next) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.filehandle.stat()];
                    case 1:
                        next = (_c.sent()).size;
                        _c.label = 2;
                    case 2:
                        blockCompressedLength = next - compressedPosition;
                        return [4 /*yield*/, this.filehandle.read(blockBuffer, 0, blockCompressedLength, compressedPosition)
                            // uncompress it
                        ];
                    case 3:
                        _c.sent();
                        return [4 /*yield*/, (0, unzip_1.unzip)(blockBuffer.slice(0, blockCompressedLength))];
                    case 4:
                        unzippedBuffer = _c.sent();
                        return [2 /*return*/, unzippedBuffer];
                }
            });
        });
    };
    BgzFilehandle.prototype.read = function (buf, offset, length, position) {
        return __awaiter(this, void 0, void 0, function () {
            var blockPositions, blockBuffer, destinationOffset, bytesRead, blockNum, uncompressedBuffer, _a, uncompressedPosition, sourceOffset, sourceEnd;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.gzi.getRelevantBlocksForRead(length, position)];
                    case 1:
                        blockPositions = _b.sent();
                        blockBuffer = Buffer.allocUnsafe(32768 * 2);
                        destinationOffset = offset;
                        bytesRead = 0;
                        blockNum = 0;
                        _b.label = 2;
                    case 2:
                        if (!(blockNum < blockPositions.length - 1)) return [3 /*break*/, 5];
                        return [4 /*yield*/, this._readAndUncompressBlock(blockBuffer, blockPositions[blockNum], blockPositions[blockNum + 1])];
                    case 3:
                        uncompressedBuffer = _b.sent();
                        _a = blockPositions[blockNum], uncompressedPosition = _a[1];
                        sourceOffset = uncompressedPosition >= position ? 0 : position - uncompressedPosition;
                        sourceEnd = Math.min(position + length, uncompressedPosition + uncompressedBuffer.length) - uncompressedPosition;
                        if (sourceOffset >= 0 && sourceOffset < uncompressedBuffer.length) {
                            uncompressedBuffer.copy(buf, destinationOffset, sourceOffset, sourceEnd);
                            destinationOffset += sourceEnd - sourceOffset;
                            bytesRead += sourceEnd - sourceOffset;
                        }
                        _b.label = 4;
                    case 4:
                        blockNum += 1;
                        return [3 /*break*/, 2];
                    case 5: return [2 /*return*/, { bytesRead: bytesRead, buffer: buf }];
                }
            });
        });
    };
    return BgzFilehandle;
}());
exports["default"] = BgzFilehandle;
//# sourceMappingURL=bgzFilehandle.js.map