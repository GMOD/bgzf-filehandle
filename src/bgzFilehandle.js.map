{"version":3,"file":"bgzFilehandle.js","sourceRoot":"","sources":["bgzFilehandle.ts"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,iCAA+B;AAC/B,wDAAiC;AACjC,yDAAiE;AAEjE;IAIE,uBAAY,EAUX;YATC,UAAU,gBAAA,EACV,IAAI,UAAA,EACJ,aAAa,mBAAA,EACb,OAAO,aAAA;QAOP,IAAI,UAAU;YAAE,IAAI,CAAC,UAAU,GAAG,UAAU,CAAA;aACvC,IAAI,IAAI;YAAE,IAAI,CAAC,UAAU,GAAG,IAAI,8BAAS,CAAC,IAAI,CAAC,CAAA;;YAC/C,MAAM,IAAI,SAAS,CAAC,2CAA2C,CAAC,CAAA;QAErE,IAAI,CAAC,aAAa,IAAI,CAAC,OAAO,IAAI,CAAC,IAAI;YACrC,MAAM,IAAI,SAAS,CAAC,iDAAiD,CAAC,CAAA;QAExE,IAAI,CAAC,GAAG,GAAG,IAAI,qBAAQ,CAAC;YACtB,UAAU,EAAE,aAAa;YACzB,IAAI,EAAE,CAAC,aAAa,IAAI,CAAC,OAAO,IAAI,IAAI,CAAC,CAAC,CAAC,OAAO,CAAC,CAAC,CAAI,IAAI,SAAM;SACnE,CAAC,CAAA;IACJ,CAAC;IAEK,4BAAI,GAAV;;;;;;4BACyB,qBAAM,IAAI,CAAC,UAAU,CAAC,IAAI,EAAE,EAAA;;wBAA7C,cAAc,GAAG,SAA4B;wBAC5C,KAAA,CAAA,KAAA,MAAM,CAAA,CAAC,MAAM,CAAA;8BAAC,cAAc;;wBAC3B,qBAAM,IAAI,CAAC,uBAAuB,EAAE,EAAA;4BAD5C,sBAAO,yBACL,OAAI,GAAE,SAAoC;gCAC1C,SAAM,GAAE,SAAS;gCACjB,UAAO,GAAE,SAAS;sCAClB,EAAA;;;;KACH;IAEK,+CAAuB,GAA7B;;;;;4BAGmC,qBAAM,IAAI,CAAC,GAAG,CAAC,YAAY,EAAE,EAAA;;wBAAxD,KAA2B,SAA6B,EAArD,oBAAoB,QAAA;wBAEZ,qBAAM,IAAI,CAAC,UAAU,CAAC,IAAI,EAAE,EAAA;;wBAArC,IAAI,GAAK,CAAA,SAA4B,CAAA,KAAjC;wBAEN,GAAG,GAAG,MAAM,CAAC,WAAW,CAAC,CAAC,CAAC,CAAA;wBAGX,qBAAM,IAAI,CAAC,UAAU,CAAC,IAAI,CAAC,GAAG,EAAE,CAAC,EAAE,CAAC,EAAE,IAAI,GAAG,EAAE,GAAG,CAAC,CAAC,EAAA;;wBAAlE,SAAS,GAAK,CAAA,SAAoD,CAAA,UAAzD;wBACjB,IAAI,SAAS,KAAK,CAAC;4BAAE,MAAM,IAAI,KAAK,CAAC,YAAY,CAAC,CAAA;wBAC5C,yBAAyB,GAAG,GAAG,CAAC,YAAY,CAAC,CAAC,CAAC,CAAA;wBACrD,sBAAO,oBAAoB,GAAG,yBAAyB,EAAA;;;;KACxD;IAEK,+CAAuB,GAA7B,UACE,WAAmB,EACnB,EAA8B,EAC9B,EAAkC;YADjC,kBAAkB,QAAA;YAClB,sBAAsB,QAAA;;;;;;wBAEnB,IAAI,GAAG,sBAAsB,CAAA;6BAC7B,CAAC,IAAI,EAAL,wBAAK;wBACC,qBAAM,IAAI,CAAC,UAAU,CAAC,IAAI,EAAE,EAAA;;wBAApC,IAAI,GAAG,CAAC,SAA4B,CAAC,CAAC,IAAI,CAAA;;;wBAItC,qBAAqB,GAAG,IAAI,GAAG,kBAAkB,CAAA;wBAEvD,qBAAM,IAAI,CAAC,UAAU,CAAC,IAAI,CACxB,WAAW,EACX,CAAC,EACD,qBAAqB,EACrB,kBAAkB,CACnB;4BAED,gBAAgB;0BAFf;;wBALD,SAKC,CAAA;wBAGsB,qBAAM,IAAA,aAAK,EAChC,WAAW,CAAC,KAAK,CAAC,CAAC,EAAE,qBAAqB,CAAC,CAC5C,EAAA;;wBAFK,cAAc,GAAG,SAEtB;wBAED,sBAAO,cAAwB,EAAA;;;;KAChC;IAEK,4BAAI,GAAV,UAAW,GAAW,EAAE,MAAc,EAAE,MAAc,EAAE,QAAgB;;;;;4BAE/C,qBAAM,IAAI,CAAC,GAAG,CAAC,wBAAwB,CAC5D,MAAM,EACN,QAAQ,CACT,EAAA;;wBAHK,cAAc,GAAG,SAGtB;wBACK,WAAW,GAAG,MAAM,CAAC,WAAW,CAAC,KAAK,GAAG,CAAC,CAAC,CAAA;wBAE7C,iBAAiB,GAAG,MAAM,CAAA;wBAC1B,SAAS,GAAG,CAAC,CAAA;wBAEX,QAAQ,GAAG,CAAC;;;6BAChB,CAAA,QAAQ,GAAG,cAAc,CAAC,MAAM,GAAG,CAAC,CAAA;wBAIT,qBAAM,IAAI,CAAC,uBAAuB,CAC3D,WAAW,EACX,cAAc,CAAC,QAAQ,CAAC,EACxB,cAAc,CAAC,QAAQ,GAAG,CAAC,CAAC,CAC7B,EAAA;;wBAJK,kBAAkB,GAAG,SAI1B;wBACK,KAA2B,cAAc,CAAC,QAAQ,CAAC,EAAhD,oBAAoB,QAAA,CAA4B;wBACnD,YAAY,GAChB,oBAAoB,IAAI,QAAQ,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,QAAQ,GAAG,oBAAoB,CAAA;wBAClE,SAAS,GACb,IAAI,CAAC,GAAG,CACN,QAAQ,GAAG,MAAM,EACjB,oBAAoB,GAAG,kBAAkB,CAAC,MAAM,CACjD,GAAG,oBAAoB,CAAA;wBAC1B,IAAI,YAAY,IAAI,CAAC,IAAI,YAAY,GAAG,kBAAkB,CAAC,MAAM,EAAE;4BACjE,kBAAkB,CAAC,IAAI,CAAC,GAAG,EAAE,iBAAiB,EAAE,YAAY,EAAE,SAAS,CAAC,CAAA;4BACxE,iBAAiB,IAAI,SAAS,GAAG,YAAY,CAAA;4BAC7C,SAAS,IAAI,SAAS,GAAG,YAAY,CAAA;yBACtC;;;wBApBD,QAAQ,IAAI,CAAC,CAAA;;4BAuBf,sBAAO,EAAE,SAAS,WAAA,EAAE,MAAM,EAAE,GAAG,EAAE,EAAA;;;;KAClC;IACH,oBAAC;AAAD,CAAC,AAvHD,IAuHC"}