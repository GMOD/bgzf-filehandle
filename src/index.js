"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.unzipChunkSlice = exports.unzipChunk = exports.unzip = exports.BgzfFilehandle = void 0;
var bgzFilehandle_1 = __importDefault(require("./bgzFilehandle"));
exports.BgzfFilehandle = bgzFilehandle_1["default"];
var unzip_1 = require("./unzip");
exports.unzip = unzip_1.unzip;
exports.unzipChunk = unzip_1.unzipChunk;
exports.unzipChunkSlice = unzip_1.unzipChunkSlice;
//# sourceMappingURL=index.js.map