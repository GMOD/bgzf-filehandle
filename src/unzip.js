const zlib = require('zlib')
const { promisify } = require('es6-promisify')

const gunzip = promisify(zlib.gunzip)

const { Z_SYNC_FLUSH, Inflate } = require('pako')

// browserify-zlib, which is the zlib shim used by default in webpacked code,
// does not properly uncompress bgzf chunks that contain more than
// one bgzf block, so export an unzip function that uses pako directly
// if we are running in a browser.
async function pakoUnzip(inputData) {
  let strm
  let pos = 0
  let i = 0
  const chunks = []
  let inflator
  do {
    const remainingInput = inputData.slice(pos)
    inflator = new Inflate()
    ;({ strm } = inflator)
    inflator.push(remainingInput, Z_SYNC_FLUSH)
    if (inflator.err) throw new Error(inflator.msg)

    pos += strm.next_in
    chunks[i] = Buffer.from(inflator.result)
    i += 1
  } while (strm.avail_in)

  const result = Buffer.concat(chunks)
  return result
}

// similar to pakounzip, except it does extra counting
// to return the positions of compressed and decompressed
// data offsets
async function unzipChunk(inputData) {
  let strm
  let cpos = 0
  let dpos = 0
  const blocks = []
  const cpositions = []
  const dpositions = []
  do {
    const remainingInput = inputData.slice(cpos)
    const inflator = new Inflate()
    // @ts-ignore
    ;({ strm } = inflator)
    // @ts-ignore
    inflator.push(remainingInput, Z_SYNC_FLUSH)
    if (inflator.err) throw new Error(inflator.msg)

    // @ts-ignore
    const buffer = Buffer.from(inflator.result)
    blocks.push(buffer)

    cpositions.push(cpos)
    dpositions.push(dpos)

    cpos += strm.next_in
    dpos += buffer.length
  } while (strm.avail_in)

  const buffer = Buffer.concat(blocks)
  return { buffer, cpositions, dpositions }
}

// similar to unzipChunk above but slices (0,minv.dataPosition) and (maxv.dataPosition,end) off
async function unzipChunkSlice(inputData, chunk) {
  let strm
  let cpos = 0
  let dpos = 0
  const decompressedBlocks = []
  const cpositions = []
  const dpositions = []
  do {
    const remainingInput = inputData.slice(cpos)
    const inflator = new Inflate()
    // @ts-ignore
    ;({ strm } = inflator)
    // @ts-ignore
    inflator.push(remainingInput, Z_SYNC_FLUSH)
    if (inflator.err) throw new Error(inflator.msg)

    // @ts-ignore
    const buffer = Buffer.from(inflator.result)
    decompressedBlocks.push(buffer)

    if (decompressedBlocks.length === 1 && chunk.minv.dataPosition) {
      // this is the first chunk, trim it
      decompressedBlocks[0] = decompressedBlocks[0].slice(
        chunk.minv.dataPosition,
      )
    }
    const origCpos = cpos
    cpos += strm.next_in
    dpos += buffer.length
    cpositions.push(cpos)
    dpositions.push(dpos)
    if (chunk.minv.blockPosition + origCpos >= chunk.maxv.blockPosition) {
      // this is the last chunk, trim it and stop decompressing
      // note if it is the same block is minv it subtracts that already
      // trimmed part of the slice length

      decompressedBlocks[decompressedBlocks.length - 1] = decompressedBlocks[
        decompressedBlocks.length - 1
      ].slice(
        0,
        chunk.maxv.blockPosition === chunk.minv.blockPosition
          ? chunk.maxv.dataPosition - chunk.minv.dataPosition + 1
          : chunk.maxv.dataPosition + 1,
      )

      break
    }
  } while (strm.avail_in)

  const buffer = Buffer.concat(decompressedBlocks)
  return { buffer, cpositions, dpositions }
}

// in node, just use the native unzipping with Z_SYNC_FLUSH
function nodeUnzip(input) {
  return gunzip(input, {
    finishFlush: (zlib.constants || zlib).Z_SYNC_FLUSH,
  })
}

module.exports = {
  unzip: typeof __webpack_require__ === 'function' ? pakoUnzip : nodeUnzip, // eslint-disable-line
  unzipChunk,
  nodeUnzip,
  pakoUnzip,
}
