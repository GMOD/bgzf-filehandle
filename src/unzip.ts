import zlib from 'zlib'
import { promisify } from 'es6-promisify'
import { pakoUnzip, unzipChunk, unzipChunkSlice } from './unzip-pako'

const gunzip = promisify(zlib.gunzip)

// in node, just use the native unzipping with Z_SYNC_FLUSH
function nodeUnzip(input: Buffer): Promise<Buffer> {
  //@ts-ignore
  return gunzip(input, {
    finishFlush: (zlib.constants || zlib).Z_SYNC_FLUSH,
  })
}

export { nodeUnzip as unzip, unzipChunk, unzipChunkSlice, nodeUnzip, pakoUnzip }
