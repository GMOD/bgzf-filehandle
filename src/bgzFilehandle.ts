import { unzip } from './unzip'
import GziIndex from './gziIndex'
import { LocalFile, GenericFilehandle } from 'generic-filehandle'

export default class BgzFilehandle {
  filehandle: GenericFilehandle
  gzi: GziIndex

  constructor({
    filehandle,
    path,
    gziFilehandle,
    gziPath,
  }: {
    filehandle?: GenericFilehandle
    path?: string
    gziFilehandle?: GenericFilehandle
    gziPath?: string
  }) {
    if (filehandle) {
      this.filehandle = filehandle
    } else if (path) {
      this.filehandle = new LocalFile(path)
    } else {
      throw new TypeError('either filehandle or path must be defined')
    }

    if (!gziFilehandle && !gziPath && !path) {
      throw new TypeError('either gziFilehandle or gziPath must be defined')
    }

    this.gzi = new GziIndex({
      filehandle: gziFilehandle,
      path: !gziFilehandle && !gziPath && path ? gziPath : `${path}.gzi`,
    })
  }

  async stat() {
    const compressedStat = await this.filehandle.stat()
    return Object.assign(compressedStat, {
      size: await this.getUncompressedFileSize(),
      blocks: undefined,
      blksize: undefined,
    })
  }

  async getUncompressedFileSize() {
    // read the last block's ISIZE (see gzip RFC),
    // and add it to its uncompressedPosition
    const [, uncompressedPosition] = await this.gzi.getLastBlock()

    const { size } = await this.filehandle.stat()

    const buf = Buffer.allocUnsafe(4)
    // note: there should be a 28-byte EOF marker (an empty block) at
    // the end of the file, so we skip backward past that
    const { bytesRead } = await this.filehandle.read(buf, 0, 4, size - 28 - 4)
    if (bytesRead !== 4) {
      throw new Error('read error')
    }
    const lastBlockUncompressedSize = buf.readUInt32LE(0)
    return uncompressedPosition + lastBlockUncompressedSize
  }

  async _readAndUncompressBlock(
    blockBuffer: Buffer,
    [compressedPosition]: [number],
    [nextCompressedPosition]: [number],
  ) {
    let next = nextCompressedPosition
    if (!next) {
      next = (await this.filehandle.stat()).size
    }

    // read the compressed data into the block buffer
    const blockCompressedLength = next - compressedPosition

    await this.filehandle.read(
      blockBuffer,
      0,
      blockCompressedLength,
      compressedPosition,
    )

    // uncompress it
    const unzippedBuffer = await unzip(
      blockBuffer.slice(0, blockCompressedLength),
    )

    return unzippedBuffer as Buffer
  }

  async read(buf: Buffer, offset: number, length: number, position: number) {
    // get the block positions for this read
    const blockPositions = await this.gzi.getRelevantBlocksForRead(
      length,
      position,
    )
    const blockBuffer = Buffer.allocUnsafe(32768 * 2)
    // uncompress the blocks and read from them one at a time to keep memory usage down
    let destinationOffset = offset
    let bytesRead = 0
    for (
      let blockNum = 0;
      blockNum < blockPositions.length - 1;
      blockNum += 1
    ) {
      // eslint-disable-next-line no-await-in-loop
      const uncompressedBuffer = await this._readAndUncompressBlock(
        blockBuffer,
        blockPositions[blockNum],
        blockPositions[blockNum + 1],
      )
      const [, uncompressedPosition] = blockPositions[blockNum]
      const sourceOffset =
        uncompressedPosition >= position ? 0 : position - uncompressedPosition
      const sourceEnd =
        Math.min(
          position + length,
          uncompressedPosition + uncompressedBuffer.length,
        ) - uncompressedPosition
      if (sourceOffset >= 0 && sourceOffset < uncompressedBuffer.length) {
        uncompressedBuffer.copy(buf, destinationOffset, sourceOffset, sourceEnd)
        destinationOffset += sourceEnd - sourceOffset
        bytesRead += sourceEnd - sourceOffset
      }
    }

    return { bytesRead, buffer: buf }
  }
}
