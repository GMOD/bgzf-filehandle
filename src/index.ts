import BgzfFilehandle from './bgzFilehandle'
import { unzip, unzipChunk, unzipChunkSlice } from './unzip'

export { BgzfFilehandle, unzip, unzipChunk, unzipChunkSlice }
