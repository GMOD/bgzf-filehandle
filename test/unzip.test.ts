import fs from 'fs'
import { describe, it, expect } from 'vitest'
import { unzip, unzipChunk, unzipChunkSlice } from '../src/unzip'

describe('unzip', () => {
  it('can unzip bgzip-1.txt.gz', async () => {
    const testData = fs.readFileSync(require.resolve('./data/bgzip-1.txt.gz'))
    const fromPako = await unzip(testData)
    expect(fromPako.length).toEqual(65569)
  })

  it('test error message modification', async () => {
    const testData = fs.readFileSync(require.resolve('./data/bgzip-1.txt.gz'))

    await expect(unzip(testData.slice(2))).rejects.toThrow(
      /problem decompressing block/,
    )
  })
})

describe('unzipChunk', () => {
  it('can unzip bgzip-1.txt.gz', async () => {
    const testData = fs.readFileSync(require.resolve('./data/paired.bam'))
    const { dpositions, cpositions } = await unzipChunk(testData)
    expect(dpositions).toMatchSnapshot()
    expect(cpositions).toMatchSnapshot()
  })
  it('test error message modification', async () => {
    const testData = fs.readFileSync(require.resolve('./data/paired.bam'))

    await expect(unzipChunk(testData.slice(2))).rejects.toThrow(
      /problem decompressing block/,
    )
  })
})

describe('unzipChunkSlice', () => {
  it('can unzip bgzip-1.txt.gz', async () => {
    const testData = fs.readFileSync(require.resolve('./data/paired.bam'))
    const { dpositions, cpositions } = await unzipChunkSlice(testData, {
      minv: { dataPosition: 0, blockPosition: 0 },
      maxv: { dataPosition: 100, blockPosition: 0 },
    })
    expect(dpositions).toMatchSnapshot()
    expect(cpositions).toMatchSnapshot()
  })
  it('can unzip bgzip-1.txt.gz positive minv.dataPosition', async () => {
    const testData = fs.readFileSync(require.resolve('./data/paired.bam'))
    const { dpositions, cpositions } = await unzipChunkSlice(testData, {
      minv: { dataPosition: 50, blockPosition: 0 },
      maxv: { dataPosition: 100, blockPosition: 0 },
    })
    expect(dpositions).toMatchSnapshot()
    expect(cpositions).toMatchSnapshot()
  })
  it('can unzip bgzip-1.txt.gz positive maxv.blockPosition', async () => {
    const testData = fs.readFileSync(require.resolve('./data/paired.bam'))
    const { dpositions, cpositions } = await unzipChunkSlice(testData, {
      minv: { dataPosition: 50, blockPosition: 0 },
      maxv: { dataPosition: 100, blockPosition: 1 },
    })
    expect(dpositions).toMatchSnapshot()
    expect(cpositions).toMatchSnapshot()
  })

  it('test error message modification', async () => {
    const testData = fs.readFileSync(require.resolve('./data/paired.bam'))
    await expect(
      unzipChunkSlice(testData.slice(2), {
        minv: { dataPosition: 40, blockPosition: 0 },
        maxv: { dataPosition: 100, blockPosition: 0 },
      }),
    ).rejects.toThrow(/problem decompressing block/)
  })
})
