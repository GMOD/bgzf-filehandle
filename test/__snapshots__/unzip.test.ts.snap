// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unzipChunk can unzip bgzip-1.txt.gz 1`] = `
Array [
  0,
  22579,
  87646,
  152675,
  217701,
  282681,
  287762,
]
`;

exports[`unzipChunk can unzip bgzip-1.txt.gz 2`] = `
Array [
  0,
  4682,
  24403,
  43638,
  62447,
  82222,
  84033,
]
`;

exports[`unzipChunkSlice can unzip bgzip-1.txt.gz 1`] = `
Array [
  0,
  22579,
]
`;

exports[`unzipChunkSlice can unzip bgzip-1.txt.gz 2`] = `
Array [
  0,
  4682,
]
`;

exports[`unzipChunkSlice can unzip bgzip-1.txt.gz positive maxv.blockPosition 1`] = `
Array [
  50,
  22579,
  87646,
]
`;

exports[`unzipChunkSlice can unzip bgzip-1.txt.gz positive maxv.blockPosition 2`] = `
Array [
  0,
  4682,
  24403,
]
`;

exports[`unzipChunkSlice can unzip bgzip-1.txt.gz positive minv.dataPosition 1`] = `
Array [
  50,
  22579,
]
`;

exports[`unzipChunkSlice can unzip bgzip-1.txt.gz positive minv.dataPosition 2`] = `
Array [
  0,
  4682,
]
`;
